libtenzir/src/tql2/set.cpp

```diff
@@ -240,16 +308,48 @@ auto set_operator::operator()(generator<table_slice> input,
       co_yield {};
       continue;
     }
-    // 1. Evaluate every right-hand side with the original input.
-    // 2. Evaluate every left-hand side as l-value and then assign.
-    // => Left side is evaluated after side effects, in order!
-    // set foo={bar: 42}, foo.bar=foo.bar+42
-    auto result = slice;
+    if (assignments_.empty()) {
+      co_yield std::move(slice);
+      continue;
+    }
+    // The right-hand side is always evaluated with the original input, because
+    // side-effects from preceding assignments shall not be reflected when
+    // calculating the value of the left-hand side.
+    auto values = std::vector<multi_series>{};
     for (auto& assignment : assignments_) {
-      auto right = eval(assignment.right, slice, ctrl.diagnostics());
-      result = assign(assignment.left, right, result, ctrl.diagnostics());
+      values.push_back(eval(assignment.right, slice, ctrl.diagnostics()));
+    }
+    // After we know all the multi series values on the right, we can split the
+    // input table slice and perform the actual assignment.
+    auto begin = int64_t{0};
+    for (auto values_slice : split_multi_series(values)) {
```

Comment:
Did you consider pushing ordering assumptions into this operator (or perhaps more generally into the new abstractions)? I assume this could make a pretty big difference in performance.